### ISSUE TITLE
There is no way to write unit test in DDlog

### ISSUE THREAD

**Issue by mbudiu-vmw** [vmware/differential-datalog#991]
It would be nice to have some form of "assert".
Also, the ability to run DDlog programs without the cli.

**Comment by ryzhyk** [comment]
Can you be more specific? We can add `panic` as a library function and build `assert` on top.  I am not convinced this is the best way to build unit tests in DDlog, so perhaps you are thinking of something else.

How would you like to run programs instead of CLI?

**Comment by mbudiu-vmw** [comment]
It would be nice to have something like rust `#[test]`in DDlog programs.
So if you have a standalone DDlog program with relations that have a static contents you can write tests against that.
```
R(0).
R(1).
O(x) :- R(x), not R(x + 1).
E(0).
AssertEquals(E, O).
AssertSubset(E, O).
```
The nice thing is that these could be compiled into a standalone test executable and run without any other tools being involved (no CLI needed, and no dat files needed).

**Comment by ryzhyk** [comment]
> It would be nice to have something like rust #[test]in DDlog programs.

Makes sense.  This way you can have tests inside individual modules that will all execute when compiling the program with `--test` flag.  Tests for modules in the `lib` directory, which now live in a separate location, could then be mostly moved to library modules themselves.

> AssertEquals(E, O).

Makes sense, but there are going to be subtle issues, e.g., in your example, `E` and `O` don't have the same type.  Also, I'm not sure if this is a new language construct (i.e., it's not a type, function, relation, rule or transformer) or perhaps it can be framed as a transformer.

One other type of assertion would be an attribute that indicates that a relation must be empty; in other words any value in the relation represents a test failure.

```
#[must_be_empty]
relation Error(x)
Error("${e} is in E but not O") :- E(e), not O(e).
Error("${o} is in O but not E) :- O(o), not E(o).
```

> The nice thing is that these could be compiled into a standalone test executable and run without any other tools being involved (no CLI needed, and no dat files needed).

I don't think this requires a new compilation mode. If unit tests don't require input data, then you don't need to provide a dat file.

On the other hand, you may want to write a test consisting of multiple transactions (e.g., if the test involves streams), in which case being able to provide a dat file may help.

**Comment by mbudiu-vmw** [comment]
Relation equality can be converted by the compiler to empty relation using a symmetric difference.
So probably empty relation is a good primitive to have. 
I am not arguing that you won't need tests using the CLI, in fact having there an assert may be very useful. But the CLI today can't even capture the output from relations -- except dump it. So enhancing the CLI language is the next step.
